File: c/in_time.c
#include <stdio.h>
#include <stdlib.h>

#define IN_TIME(time, from, to) (\
    ( ( (from) < (to) ) && ( (from) <= (time) && (time) < (to) ) ) || \
    ( ( (from) == (to) ) && ( (from) == (time) ) ) || \
    ( ( (from) > (to) ) && ( (from) <= (time) || (time) < (to) ) ) \
)

void abort(void);
#define TEST(msg, cond) { if ( !(cond) ) { fprintf(stderr,  msg "\n"); abort();  } }

int main()
{

  TEST("before start time should not be in time.", !IN_TIME(5, 6, 20));
  TEST("start time should be in time.", IN_TIME(6, 6, 20));
  TEST("before end time should be in time.", IN_TIME(19, 6, 20));
  TEST("end time should not be in time.", !IN_TIME(20, 6, 20));

  TEST("before start time should not be in time (reverse)", !IN_TIME(19, 20, 6));
  TEST("start time should be in time ( reverse )", IN_TIME(20, 20, 6));
  TEST("before end time should be in time ( reverse )", IN_TIME(5, 20, 6));
  TEST("end time should noo be in time ( reverse ).", !IN_TIME(6, 20, 6));

  TEST("in case start time and end time.", IN_TIME(5, 5, 5));
  TEST("in case start time and end time.", !IN_TIME(4, 5, 5));

  printf("success!! all test passed!!\n");
  return 0;
}

File: perl/in_time.pl
package InTime;
use strict;
use warnings;

sub in_time {
  my ( $time , %args ) = @_;

  my ( $from, $to ) = ( $args{from}, $args{to} );

  if ( $from < $to ) {
    return 1 if ( $from <= $time && $time < $to );
  } elsif ( $from == $to ) {
    return 1 if $from == $time;
  } else {
    return 1 if ( $from <= $time || $time < $to );
  }

}

package main;
use strict;
use warnings;
use Test::More;

subtest 'normal case' => sub {
    ok(!InTime::in_time(9, from => 10, to => 20) , "before start time should not be in time.");
    ok(InTime::in_time(10, from => 10, to => 20), "start time should be in time.");
    ok(InTime::in_time(19, from => 10, to => 20), "before end time should be in time.");
    ok(!InTime::in_time(20, from => 10, to => 20), "before end time should not be in time.");
};

subtest 'reverse order case' => sub {
    ok(!InTime::in_time(19, from => 20, to => 10), "before start time should not be in time ( reverse )");
    ok(InTime::in_time(20, from => 20, to => 10), "start time should be in time ( reverse )");
    ok(InTime::in_time(9, from => 20, to => 10), "before end time should be in time ( reverse )");
    ok(!InTime::in_time(10, from => 20, to => 10), "end time should not be in time ( reverse )");
};

subtest 'in case from time equal to time' => sub {
    ok(InTime::in_time(10, from => 10, to => 10), "in case from time equal to time");
    ok(!InTime::in_time(9, from => 10, to => 10), "in case from time equal to time");
    ok(!InTime::in_time(11, from => 10, to => 10), "in case from time equal to time");
};

done_testing;
